### ISSUE TITLE
replacement has length zero bug

### ISSUE THREAD

**Issue by bwlewis** [bwlewis/irlba#7]
reported by @koheiw   moved from issue #5 thread:

``` r
mx <- matrix(sample(1:100, 10^3 * 10^3, replace=TRUE), nrow=10^3)
S <- irlba(mx, nv=300, verbose=TRUE, center=colMeans(mx), right_only=TRUE)
```

```
Working dimension size 317
Error in W[, j_w] <- W[, j_w] - ds * drop(cross(dv, VJ)) * du : 
  replacement has length zero
```

**Comment by bwlewis** [comment]
It's a dumb bug on my part that was not covered in the unit tests, sorry. For now, add `fastpath=FALSE`  until a new version is ready.

**Comment by mahermassoud** [comment]
I am getting this issue. Adding fastpath = FALSE does not do anything

**Comment by bwlewis** [comment]
@mahermassoud can you provide a reproducible example? I am not able to reproduce this.  Also, please note R.version. Thanks.

**Comment by mahermassoud** [comment]
```
> packageDescription("irlba")
Package: irlba
Type: Package
Title: Fast Truncated Singular Value Decomposition and Principal Components Analysis for Large Dense and
           Sparse Matrices
Version: 2.2.1
Date: 2017-05-16
Authors@R: c( person("Jim", "Baglama", rol=c("aut", "cph"), email="jbaglama@uri.edu"), person("Lothar",
           "Reichel", role=c("aut", "cph"), email="reichel@math.kent.edu"), person("B. W.", "Lewis",
           role=c("aut","cre","cph"), email="blewis@illposed.net"))
Description: Fast and memory efficient methods for truncated singular value decomposition and principal
           components analysis of large sparse and dense matrices.
Depends: Matrix
LinkingTo: Matrix
Imports: stats, methods
License: GPL-3
BugReports: https://github.com/bwlewis/irlba/issues
RoxygenNote: 5.0.1
NeedsCompilation: yes
Packaged: 2017-05-16 20:24:46 UTC; blewis
Author: Jim Baglama [aut, cph], Lothar Reichel [aut, cph], B. W. Lewis [aut, cre, cph]
Maintainer: B. W. Lewis <blewis@illposed.net>
Repository: CRAN
Date/Publication: 2017-05-17 07:28:01 UTC
Built: R 3.3.2; x86_64-apple-darwin13.4.0; 2017-05-18 12:04:26 UTC; unix

-- File: /Library/Frameworks/R.framework/Versions/3.3/Resources/library/irlba/Meta/package.rds 
```

**Comment by bwlewis** [comment]
ok  thanks.

i don't have a Mac  but will ask a friend to investigate.

**Comment by bwlewis** [comment]
Aha! I was able to get a bug on a different machine with a different BLAS:

```{r}
library(irlba)
mx = matrix(sample(1:100, 10^3 * 10^3, replace=TRUE), nrow=10^3)
set.seed(1)
fast = irlba(mx, nv=300, center=colMeans(mx), verbose=TRUE)

## Working dimension size 307
## Initializing starting vector v with samples from standard normal distribution.
## Use `set.seed` first for reproducibility.
## irlba: using fast C implementation
## Error in irlba(mx, nv = 300, center = colMeans(mx), verbose = TRUE) :
##  BLAS/LAPACK routine 'DLASCL' gave error code -4
```

Continuing to investigate. Using either fastpath=FALSE or right_only=TRUE (which implies fastpath=FALSE) does seem to work for me.

**Comment by mahermassoud** [comment]
```{r}
library(irlba)
mx <- matrix(sample(1:100, 10^3 * 10^3, replace=TRUE), nrow=10^3)
S <- irlba(mx, nv=300,  center=colMeans(mx), right_only=TRUE)
```

this works just fine

**Comment by bwlewis** [comment]
OK, thanks. Something is going on in the fast code path. Investigating...

**Comment by mahermassoud** [comment]
Adding fastpath = FALSE or right_only = TRUE or both gives me this
`Error in irlba::irlba(x, n_pcs, center = mu, scale = s, right_only = TRUE) : 
  starting vector near the null space`

**Comment by bwlewis** [comment]
that is a probably related issue, but yet a little different and discussed (briefly) in the documentation.  You can try a different starting vector for the iteration to get around that, either by explicitly supplying a 'v' argument, or by setting a seed before running irlba.

**Comment by bwlewis** [comment]
Arrrghh! The matrix `mx` above is an integer matrix and I'm forgetting to check for that, which is a bug.

Try

```{r}
irlba(mx + 0.0, nv=300)
```

I will fix this in the package soon...

**Comment by bwlewis** [comment]
@mahermassoud   thanks very much for your help with this, I think it might be fixed now. Can you try again with the current GitHub master branch?

**Comment by mahermassoud** [comment]
When I try to install I get the following. Not sure if this is on your end or mine

`> library(devtools)
> install_github("bwlewis/irlba")
Downloading GitHub repo bwlewis/irlba@master
from URL https://api.github.com/repos/bwlewis/irlba/zipball/master
Installing irlba
'/Library/Frameworks/R.framework/Resources/bin/R' --no-site-file --no-environ --no-save --no-restore --quiet CMD  \
  INSTALL  \
  '/private/var/folders/9m/dy7kzyp50fldn0ydz_9wj8180000gn/T/Rtmpszzlgh/devtools21b3630e5d8/bwlewis-irlba-4d0cc15'  \
  --library='/Library/Frameworks/R.framework/Versions/3.3/Resources/library' --install-tests 

* installing *source* package ‘irlba’ ...
** libs
clang -I/Library/Frameworks/R.framework/Resources/include -DNDEBUG  -I/usr/local/include -I/usr/local/include/freetype2 -I/opt/X11/include -I"/Library/Frameworks/R.framework/Versions/3.3/Resources/library/Matrix/include"   -fPIC  -Wall -mtune=core2 -g -O2  -c irlb.c -o irlb.o
clang -I/Library/Frameworks/R.framework/Resources/include -DNDEBUG  -I/usr/local/include -I/usr/local/include/freetype2 -I/opt/X11/include -I"/Library/Frameworks/R.framework/Versions/3.3/Resources/library/Matrix/include"   -fPIC  -Wall -mtune=core2 -g -O2  -c utility.c -o utility.o
clang -dynamiclib -Wl,-headerpad_max_install_names -undefined dynamic_lookup -single_module -multiply_defined suppress -L/Library/Frameworks/R.framework/Resources/lib -L/usr/local/lib -o irlba.so irlb.o utility.o -L/Library/Frameworks/R.framework/Resources/lib -lRlapack -L/Library/Frameworks/R.framework/Resources/lib -lRblas -L/usr/local/lib/gcc/x86_64-apple-darwin13.0.0/4.8.2 -lgfortran -lquadmath -lm -F/Library/Frameworks/R.framework/.. -framework R -Wl,-framework -Wl,CoreFoundation
ld: warning: directory not found for option '-L/usr/local/lib/gcc/x86_64-apple-darwin13.0.0/4.8.2'
ld: library not found for -lgfortran
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [irlba.so] Error 1
ERROR: compilation failed for package ‘irlba’
* removing ‘/Library/Frameworks/R.framework/Versions/3.3/Resources/library/irlba’
* restoring previous ‘/Library/Frameworks/R.framework/Versions/3.3/Resources/library/irlba’
Installation failed: Command failed (1)`

**Comment by bwlewis** [comment]
It looks like you're missing some R required Fortran compiler tools on your Mac. I don't know too much about Macs though, so I will be of limited help.

I am pretty sure this issue is fixed and will appear on a CRAN update in a month or so.

You can manually get around the problem by avoiding integer matrices, for example instead of `irlba(mx, nv=300)`  use `irlba(mx + 0.0, nv=300)` in the example above.

**Comment by anachshon** [comment]
I had the same problem, and i believe i managed to install the version 2.2.2 from the GitHub and it still did not solve the problem. I am running R version 3.3.2 on a mac os machine.
The trick of adding +0.0 did it which is nice :-). 
Thanks for this useful workaround.

**Comment by bwlewis** [comment]
This should be fixed now as of the master branch version 2.2.2, see https://github.com/bwlewis/irlba/blob/master/R/irlba.R#L257

If not, pls re-open.
